pkg/registry/core/service/ipallocator/ipallocator.go

```diff
@@ -0,0 +1,480 @@
+	if err != nil {
+		return net.IP{}, err
+	}
+	if a.Free() == 0 {
+		a.metrics.incrementAllocationErrors(label.String(), "dynamic")
+		return net.IP{}, ErrFull
+	}
+
+	// size of the upper subrange, prioritized for random allocation
+	subrangeMax := a.max - int64(a.offset)
+	// try to get a value from the upper range [a.offset, max]
+	start := a.rand.Int63n(subrangeMax) + int64(a.offset)
+	ip := getIndexedAddr(prefix, start)
+	offsetAddress := getIndexedAddr(prefix, int64(a.offset))
+	var i int64
+	for i = 0; i < subrangeMax; i++ {
+		ip = ip.Next()
+		// roll over if it is greater than the last IP allocatable in the range
+		if ip.Compare(a.lastAddress) == 1 {
+			ip = offsetAddress.Next()
+		}
+		name := ip.String()
+		_, err = a.ipAddressLister.Get(name)
+		if !apierrors.IsNotFound(err) {
+			continue
+		}
+		err = a.createIPAddress(name, svc, "dynamic")
+		// an error can happen if there is a race and our informer was not updated
+		// swallow the error and try with the next IP address
+		if err == nil {
+			return ip.AsSlice(), nil
+		}
+	}
+
+	// rand.Int63n panics for values <= 0
+	if a.offset == 0 {
+		a.metrics.incrementAllocationErrors(label.String(), "dynamic")
+		return net.IP{}, ErrFull
+	}
+
+	// try to get a value from the lower range [0, a.offset]
+	start = a.rand.Int63n(int64(a.offset))
+	ip = getIndexedAddr(prefix, start)
+	for i = 0; i < int64(a.offset); i++ {
+		ip = ip.Next()
+		// roll over if it is greater than the last IP allocatable in the range
+		if ip.Less(a.firstAddress) || ip.Compare(offsetAddress) == 1 {
+			ip = a.firstAddress
+		}
+		name := ip.String()
+		_, err = a.ipAddressLister.Get(name)
+		if !apierrors.IsNotFound(err) {
+			continue
+		}
+		err = a.createIPAddress(name, svc, "dynamic")
+		// an error can happen if there is a race and our informer was not updated
+		// swallow the error and try with the next IP address
+		if err == nil {
+			return ip.AsSlice(), nil
+		}
+	}
+
+	// update metrics
+	a.metrics.incrementAllocationErrors(label.String(), "dynamic")
+	return net.IP{}, ErrFull
+
+}
+
+// Release releases the IP back to the pool. Releasing an
+// unallocated IP or an IP out of the range is a no-op and
+// returns no error.
+func (a *Allocator) Release(ip net.IP) error {
+	return a.release(ip, dryRunFalse)
+}
+
+func (a *Allocator) release(ip net.IP, dryRun bool) error {
+	if !a.ipAddressSynced() {
+		return fmt.Errorf("allocator not ready")
+	}
+	if dryRun {
+		return nil
+	}
+	name := ip.String()
+	if _, err := a.ipAddressLister.Get(name); err != nil {
+		return nil
+	}
+	// convert IP to name
+	err := a.client.IPAddresses().Delete(context.Background(), name, metav1.DeleteOptions{})
+	if err == nil {
+		// update metrics
+		label := a.CIDR()
+		a.metrics.setAllocated(label.String(), a.Used())
+		a.metrics.setAvailable(label.String(), a.Free())
+	}
+	return nil
+}
+
+// ForEach executes the function in each allocated IP
+func (a *Allocator) ForEach(f func(net.IP)) {
```

Comment:
why do we need this func?

Reply:
It is used by the Repair loop on the bitmap allocator.
This new allocator don't really need it but it needs to match the Interface so we can switch the old one with this